File: Strivers-A2Z-DSA-Sheet/05.Recursion/2.Subsequences Pattern/javascript/09.Combination Sum 3.js
/*
=============================================================================
  QUESTION: 216. Combination Sum III (LeetCode)
=============================================================================

  Find all combinations of k numbers from 1-9 (each used once) summing to n.

  Examples:
    k=3, n=7 → [[1,2,4]]
    k=3, n=9 → [[1,2,6],[1,3,5],[2,3,4]]

=============================================================================
  APPROACH: Pick / Not-Pick (1 to 9) — O(2^9) Time, O(k) Space
=============================================================================

  Same as Combination Sum 2 but with fixed pool: numbers 1 to 9.
  Extra constraint: exactly k numbers must be picked.

  Dry Run k=3, n=7:
    pick 1 → pick 2 → pick 3: sum=6≠7 ✗ | pick 4: sum=7 & k=3 → PUSH [1,2,4] ✓
                     → skip 2 → pick 3 → pick 4: sum=8>7 ✗
    skip 1 → pick 2 → pick 3 → pick 4: sum=9>7 ✗
    Result: [[1,2,4]] ✓

=============================================================================
*/

function combinationSum3(k, n) {
    let result = [];

    function solve(num, target, temp) {
        // Found valid combination: exactly k numbers summing to target
        if (target === 0 && temp.length === k) {
            result.push([...temp]);
            return;
        }

        // Invalid: out of range or exceeded target
        if (num > 9 || target < 0) return;

        // PICK current number
        if (num <= target) {
            temp.push(num);
            solve(num + 1, target - num, temp);
            temp.pop();  // Backtrack
        }

        // SKIP current number
        solve(num + 1, target, temp);
    }

    solve(1, n, []);
    return result;
}

// ==========================================================================
// DRIVER CODE
// ==========================================================================
console.log(combinationSum3(3, 7));   // [[1,2,4]]
console.log(combinationSum3(3, 9));   // [[1,2,6],[1,3,5],[2,3,4]]
console.log(combinationSum3(2, 5));   // [[1,4],[2,3]]

File: Strivers-A2Z-DSA-Sheet/05.Recursion/2.Subsequences Pattern/python/09.Combination Sum 3.py
"""
=============================================================================
  QUESTION: 216. Combination Sum III (LeetCode)
=============================================================================

  Find all combinations of k numbers (1-9, each at most once) summing to n.
  Example: k=3, n=7 → [[1,2,4]] (1+2+4=7)

  APPROACH: Pick/Not-Pick with numbers 1-9.
  Extra constraint: exactly k numbers must be picked.
  Base: target == 0 AND len(temp) == k → valid.

  DRY RUN k=3, n=7:
    Pick 1 → Pick 2 → Pick 3(sum=6≠7) → Skip 3 → Pick 4(sum=7, k=3) → [1,2,4] ✓
    All other combos either exceed 7 or have wrong count.

  Time: O(2^9), Space: O(k)
=============================================================================
"""


def solve(i, k, target, temp, ans):
    if target == 0 and len(temp) == k:
        ans.append(temp[:])
        return
    if i > 9 or target < 0:
        return

    if i <= target:
        temp.append(i)
        solve(i + 1, k, target - i, temp, ans)
        temp.pop()

    solve(i + 1, k, target, temp, ans)


def combination_sum3(k, n):
    ans = []
    solve(1, k, n, [], ans)
    return ans


# Driver Code
if __name__ == "__main__":
    print(combination_sum3(3, 7))   # [[1,2,4]]
    print(combination_sum3(3, 9))   # [[1,2,6],[1,3,5],[2,3,4]]

